### DISCUSSION TITLE
Xilinx UltraScale+ MPSoC A53 core/SMP support?

### DISCUSSION THREAD

**Discussion by rowol** [OpenAMP/open-amp#422] [Q&A] [answered]
Hi, 

Not sure if this is an appropriate place to ask these questions, if not, please feel free to delete this post.  

We are using a Zynq Ulltrascale+ MPSoC with four A53 cores (and two R5 cores.)  We have some existing bare metal firmware running on an A53 core using Xilinx's Bare Metal Software Stack.  Because of the way this firmware accesses DDR, for high performance reasons it needs to run on an A53 core (vs an R5.)  We would like to run PetaLinux on a second A53 core.

At first glance OpenAMP sounds like a nice solution, but digging deeper, Xilinx doesn't appear to provide a remoteproc PetaLinux driver for the A53: they only provide drivers for the R5 (Zynq Ulltrascale+ MPSoC) and the A9 (Zynq 7000)  cores.

**Questions**
- Do you happen to know if anyone provides an A53 remoteproc solution?  (either free or commercial...)   
- Barring that, would it be really hard for me to write that driver?   (... and presumably I might also have to write some libmetal functions?)   _Source for the R5 and A9 drivers is available_
- Would I be better off using a hypervisor, or perhaps modifying the FSBL (or u-boot) to do I want instead?   I'm interested in the simplest solution.   We would like to be able to load firmware into the second core and start it running.   I don't think (?) we would need to be able to stop that core.   As a guess, for intercore communication, we'd probably be able to use shared memory, and possibly also mailslots or something similar.

Thanks!

**Comment by edmooring** [comment, marked as answer]
You will probably be able to get a more definitive answer from Xilinx, but here's a quick set of answers for you:

- I don't believe there is an A53 remoteproc solution out there.

- It would be rather challenging to write such a driver, since the A53 cluster shares a lot of resources between the cores and partitioning those resources would be interesting. At a rough guess, it would require modifications to FSBL, Arm Trusted Firmware, U-Boot, and the Linux kernel. 
- It would probably be easier to use a hypervisor, and bypass remoteproc completely. I think most of the infrastructure to do what you want is already present in Xen.

I hope this helps. I can't really give you a clearer answer without knowing your application.

**Reply by rowol** [reply]
Thanks, that was actually a pretty good answer.   I also posted on Xilinx's forum, they haven't responded, but I've seen older posts on there that say they only support R5 and A9 remoteproc (since we're using an FPGA with R5s and A53s, that means R5 only.)   The Xen hypervisor was one of the other solutions we looked at, and may still wind up considering. 

I think what we are going to try is using Xilinx's FSBL (first stage bootloader) to load two A53 cores from two separate ELF files in BOOT.BIN  (one will be the u-boot/rootfs deal), so we can boot PetaLinux on one and bare-metal firmware on the other.   It appears to have a sparsely documented feature that makes that work, and I think (?) it will start both cores running, although I haven't verified this yet.
